src/FlashLoanVault.sol

```diff
@@ -0,0 +1,285 @@
+    
+    /// @notice Emit event when ERC20 tokens are withdrawn from Fractionalized Wrapper
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+
+    /// @notice User to deposit underlying tokens for FractionalWrapper tokens
+    /// @dev Mints shares Vault shares to receiver by depositing exactly amount of underlying tokens
+    /// @param assets The amount of underlying tokens to deposit
+    /// @param receiver Address of receiver of Fractional Wrapper shares
+    function deposit(uint256 assets, address receiver) external returns(uint256 shares) {       
+        receiver = msg.sender;
+        shares = convertToShares(assets);
+
+        //transfer DAI from user
+        underlying.safeTransferFrom(receiver, address(this), assets);
+
+        //mint yvDAI to user
+        bool sent = _mint(receiver, shares);
+        require(sent, "Mint failed!"); 
+
+        emit Deposit(msg.sender, receiver, assets, shares);  
+    }
+
+
+    /// @dev Burns shares from owner and sends exactly assets of underlying tokens to receiver; based on the exchange rate.
+    /// @param assets The amount of underlying tokens to withdraw
+    /// @param receiver Address of receiver of underlying tokens - DAI
+    /// @param owner Address of owner of Fractional Wrapper shares - yvDAI
+    function withdraw(uint256 assets, address receiver, address owner) external returns(uint256 shares) {
+        shares = convertToShares(assets);
+        
+        // MUST support a withdraw flow where the shares are burned from owner directly where owner is msg.sender,
+        // OR msg.sender has ERC-20 approval over the shares of owner
+        if(msg.sender != owner){
+            uint allowedShares = _allowance[owner][receiver] ;
+            require(allowedShares >= shares, "Allowance exceeded!");
+            _allowance[owner][receiver] = allowedShares - shares;
+        }
+
+        //burn wrapped tokens(shares) -> yvDAI 
+        burn(owner, shares);
+        
+        //transfer assets
+        underlying.safeTransfer(receiver, assets);
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+    /// @dev Burns shares from owner and sends assets of underlying tokens to receiver; based on the exchange rate.
+    /// @param shares The amount of wrapped tokens to redeem for underlying tokens (assets)
+    /// @param receiver Address of receiver of underlying tokens - DAI
+    /// @param owner Address of owner of Fractional Wrapper shares - yvDAI
+    function redeem(uint256 shares, address receiver, address owner) external returns(uint256 assets) {
+        assets = convertToAssets(shares);
+        
+        // MUST support a redeem flow where the shares are burned from owner directly where owner is msg.sender,
+        // OR msg.sender has ERC-20 approval over the shares of owner
+        if(msg.sender != owner){
```

Comment:
Same as above. Coding an ERC-20 token from scratch is always a great exercise :D